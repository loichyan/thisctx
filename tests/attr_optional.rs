#![allow(dead_code)]

#[derive(thisctx::WithContext)]
pub enum Error {
    Optional(#[thisctx(optional = "path")] Option<String>),
    WithSource(
        #[source] String,
        #[thisctx(optional = "path")] Option<String>,
    ),
    WithFrom(
        #[thisctx(from)] String,
        #[thisctx(optional = "path")] Option<String>,
    ),
}

#[derive(thisctx::WithContext)]
#[thisctx(suffix = "Context")]
pub struct Struct {
    #[thisctx(from)]
    source: String,
    #[thisctx(optional)]
    path: Option<String>,
}
