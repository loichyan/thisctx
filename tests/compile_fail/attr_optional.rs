#![allow(dead_code)]

#[derive(thisctx::WithContext)]
enum Error {
    Optional(#[thisctx(optional = "path")] Option<String>),
    WithSource(
        #[source] String,
        #[thisctx(optional = "path")] Option<String>,
    ),
    WithFrom(
        #[thisctx(from)] String,
        #[thisctx(optional = "path")] Option<String>,
    ),
    WithConflict(
        #[thisctx(from)] String,
        #[thisctx(optional = "path")] Option<String>,
        #[thisctx(optional = "path")] Option<usize>,
    ),
}

#[derive(thisctx::WithContext)]
#[thisctx(suffix = "Context")]
struct Struct {
    #[thisctx(from)]
    source: String,
    #[thisctx(optional)]
    path: Option<String>,
    #[thisctx(optional = "path")]
    path2: Option<usize>,
}

#[derive(thisctx::WithContext)]
#[thisctx(suffix = "Context")]
struct MultipleOptionalFields {
    #[thisctx(from)]
    source: String,
    #[thisctx(optional)]
    path: Option<String>,
    #[thisctx(optional)]
    path2: Option<usize>,
}

fn main() {}
