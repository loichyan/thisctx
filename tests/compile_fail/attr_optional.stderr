error: `optional` ID conflicts: path
  --> tests/compile_fail/attr_optional.rs:14:5
   |
14 |     WithConflict(
   |     ^^^^^^^^^^^^

error: `optional` ID conflicts: path
  --> tests/compile_fail/attr_optional.rs:23:8
   |
23 | struct Struct {
   |        ^^^^^^
